foundationdb: make operational on nixos

siriobalmelli opened the pull request:
Make foundationdb usable/operational

This is a WIP PR:

- [x] update the foundationdb package to the latest release
- [I/P] fix foundationdb so it is not segfaulting
- [x] add NixOS test for foundationdb
- [x] fix NixOS foundationdb module

See #319537 and #378888

cc @PaulGrandperrin @lostnet @thoughtpolice

## Things done

WIP, will update when finished

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

siriobalmelli commented:
Running 7.3.58, I'm getting a very uninformative coredump:

```
$ coredumpctl info 78783
           PID: 78783 (fdbcli)
           UID: 1003 (sirio)
           GID: 100 (users)
        Signal: 11 (SEGV)
     Timestamp: Mon 2025-02-03 12:31:06 UTC (1min 16s ago)
  Command Line: result/bin/fdbcli --version
    Executable: /nix/store/2bjh06k648875mghamj6zn0zrr3ag758-foundationdb-7.3.58/bin/fdbcli
 Control Group: /user.slice/user-1003.slice/session-3.scope
          Unit: session-3.scope
         Slice: user-1003.slice
       Session: 3
     Owner UID: 1003 (sirio)
       Boot ID: 7a6a0994059647f181509c8e7484b9f2
    Machine ID: ec261a26cd2b681a38c69687fd7c6203
      Hostname: staging-foundation-2
       Storage: /var/lib/systemd/coredump/core.fdbcli.1003.7a6a0994059647f181509c8e7484b9f2.78783.1738585866000000.zst (present)
  Size on Disk: 281.9K
       Message: Process 78783 (fdbcli) of user 1003 dumped core.

                Stack trace of thread 78783:
                #0  0x000000000121ada4 n/a (n/a + 0x0)
                #1  0x000000000046be00 n/a (n/a + 0x0)
                #2  0x0000000000479158 n/a (n/a + 0x0)
                #3  0x000000000047cffc n/a (n/a + 0x0)
                #4  0x000000000047befc n/a (n/a + 0x0)
                #5  0x000000000047c4d4 n/a (n/a + 0x0)
                #6  0x000000000047cea4 n/a (n/a + 0x0)
                #7  0x000000000047befc n/a (n/a + 0x0)
                #8  0x000000000047bff0 n/a (n/a + 0x0)
                #9  0x000000000047c4d4 n/a (n/a + 0x0)
                #10 0x0000000000fd4f2c n/a (n/a + 0x0)
                #11 0x0000000000fd54e0 n/a (n/a + 0x0)
                #12 0x000000000045ebc4 n/a (n/a + 0x0)
                #13 0x0000ffff99a26414 n/a (n/a + 0x0)
                #14 0x0000000000461ff0 n/a (n/a + 0x0)
                ELF object binary architecture: AARCH64
```

onny commented:
does the latest release fix segmentation faults? https://github.com/apple/foundationdb/releases/tag/7.3.59

onny commented:
Unrelated for this change but I'm also looking forward to cross-compile foundationdb to riscv64. Currently fails with:

```
$ NIXPKGS_ALLOW_UNSUPPORTED_SYSTEM=1 NIXPKGS_ALLOW_BROKEN=1 nix build --impure -L .#pkgsCross.riscv64.foundationdb
[...]
-- Looking for stdatomic.h - found
-- Looking for DTRACE_PROBE
-- Looking for DTRACE_PROBE - not found
-- Looking for aligned_alloc
-- Looking for aligned_alloc - found
-- Has aligned_alloc: 1
CMake Error at /nix/store/mfrfb675a16gqchycqv16aziivn09dxa-cmake-3.31.3/share/cmake-3.31/Modules/FindPackageHandleStandardArgs.cmake:233 (message):
  Could NOT find jemalloc (missing: jemalloc_LIBRARY) (Required is at least
  version "5.3.0")
Call Stack (most recent call first):
  /nix/store/mfrfb675a16gqchycqv16aziivn09dxa-cmake-3.31.3/share/cmake-3.31/Modules/FindPackageHandleStandardArgs.cmake:603 (_FPHSA_FAILURE_MESSAGE)
  cmake/Findjemalloc.cmake:97 (find_package_handle_standard_args)
  cmake/Findjemalloc.cmake:159 (_finalize_find_package_jemalloc)
  cmake/FDBComponents.cmake:9 (find_package)
  CMakeLists.txt:72 (include)

-- Configuring incomplete, errors occurred!
```

onny commented on pkgs/servers/foundationdb/default.nix:
```diff
@@ -17,15 +17,16 @@
+    hash = "sha256-hutz65sm3D08lxc2XgfQf6+b+IAV1I27lxBAE7V5ORo=";
     boost = boost186;
     ssl = openssl;
 
```
We could also add 

```
  passthru = {
    tests.foundationdb = nixosTests.foundationdb;
  };
```

onny commented on nixos/tests/foundationdb.nix:
```diff
@@ -0,0 +1,35 @@
+      { nodes, ... }:
+      ''
+        server.wait_for_unit("foundationdb.service")
+        server.succeed("fdbcli --exec status")
```
Should we add more test cases like adding and reading stuff to the database?

siriobalmelli replied:
I agree, but for now "starting and initializing the db successfully without coredump" is the bar to clear.

siriobalmelli commented on pkgs/servers/foundationdb/cmake.nix:
```diff
@@ -2,7 +2,7 @@
 { lib, fetchFromGitHub
 , cmake, ninja, python3, openjdk, mono, pkg-config
-, msgpack-cxx, toml11, jemalloc, doctest
+, msgpack-cxx, toml11, jemalloc, doctest, zlib
```
Adding an initial nixfmt commit before changes

onny commented on pkgs/servers/foundationdb/cmake.nix:
```diff
@@ -1,143 +1,179 @@
+      meta = with lib; {
+        description = "Open source, distributed, transactional key-value store";
+        homepage = "https://www.foundationdb.org";
+        license = licenses.asl20;
```
```suggestion
        lib.license = licenses.asl20;
```

onny commented on pkgs/servers/foundationdb/cmake.nix:
```diff
@@ -1,143 +1,179 @@
+        description = "Open source, distributed, transactional key-value store";
+        homepage = "https://www.foundationdb.org";
+        license = licenses.asl20;
+        platforms = [ "x86_64-linux" ] ++ lib.optionals (!(avxEnabled version)) [ "aarch64-linux" ];
```
```suggestion
        lib.platforms = [ "x86_64-linux" ] ++ lib.optionals (!(avxEnabled version)) [ "aarch64-linux" ];
```

onny commented on pkgs/servers/foundationdb/cmake.nix:
```diff
@@ -1,143 +1,179 @@
+        platforms = [ "x86_64-linux" ] ++ lib.optionals (!(avxEnabled version)) [ "aarch64-linux" ];
+        # Fails when cross-compiling with "/bin/sh: gcc-ar: not found"
+        broken = stdenv.buildPlatform != stdenv.hostPlatform;
+        maintainers = with maintainers; [
```
```suggestion
        maintainers = with lib.maintainers; [
```

onny commented on pkgs/servers/foundationdb/cmake.nix:
```diff
@@ -1,143 +1,179 @@
+        "pythonsrc"
+      ];
+
+      meta = with lib; {
```
```suggestion
      meta = {
```

onny commented on pkgs/servers/foundationdb/default.nix:
```diff
@@ -17,15 +17,18 @@
   foundationdb73 = cmakeBuild {
-    version = "7.3.42";
-    hash = "sha256-jQcm+HLai5da2pZZ7iLdN6fpQZxf5+/kkfv9OSXQ57c=";
+    version = "7.3.62";
```
```suggestion
    version = "7.3.63";
```

siriobalmelli replied:
The odd-numbered releases have AVX enabled, we want the odd-numbered releases.

onny commented on pkgs/servers/foundationdb/default.nix:
```diff
@@ -17,15 +17,18 @@
-    version = "7.3.42";
-    hash = "sha256-jQcm+HLai5da2pZZ7iLdN6fpQZxf5+/kkfv9OSXQ57c=";
+    version = "7.3.62";
+    hash = "sha256-tPB/jJbdC11wQvPOI5KtlpQgw9yVLBhxABGgDSBpwBU=";
```
```suggestion
    hash = "sha256-fUyxV6oZdJOh0mv+uWz4hiNqyQHDR6hekTLS62XlPM8=";
```

siriobalmelli replied:
The odd-numbered releases have AVX enabled, we want the odd-numbered releases.

onny commented:
maybe we can move foundationdb into `pkgs/by-name` and get rid of the fdbPackages and `foundationdb73`-prefix?

siriobalmelli commented:
> does the latest release fix segmentation faults? https://github.com/apple/foundationdb/releases/tag/7.3.59

As of `7.3.62` I'm still seeing segfaults

> Unrelated for this change but I'm also looking forward to cross-compile foundationdb to riscv64. Currently fails with:
> 
> ```
> $ NIXPKGS_ALLOW_UNSUPPORTED_SYSTEM=1 NIXPKGS_ALLOW_BROKEN=1 nix build --impure -L .#pkgsCross.riscv64.foundationdb
> [...]
> -- Looking for stdatomic.h - found
> -- Looking for DTRACE_PROBE
> -- Looking for DTRACE_PROBE - not found
> -- Looking for aligned_alloc
> -- Looking for aligned_alloc - found
> -- Has aligned_alloc: 1
> CMake Error at /nix/store/mfrfb675a16gqchycqv16aziivn09dxa-cmake-3.31.3/share/cmake-3.31/Modules/FindPackageHandleStandardArgs.cmake:233 (message):
>   Could NOT find jemalloc (missing: jemalloc_LIBRARY) (Required is at least
>   version "5.3.0")
> Call Stack (most recent call first):
>   /nix/store/mfrfb675a16gqchycqv16aziivn09dxa-cmake-3.31.3/share/cmake-3.31/Modules/FindPackageHandleStandardArgs.cmake:603 (_FPHSA_FAILURE_MESSAGE)
>   cmake/Findjemalloc.cmake:97 (find_package_handle_standard_args)
>   cmake/Findjemalloc.cmake:159 (_finalize_find_package_jemalloc)
>   cmake/FDBComponents.cmake:9 (find_package)
>   CMakeLists.txt:72 (include)
> 
> -- Configuring incomplete, errors occurred!
> ```

This *should* work based on the fact that jemalloc is in buildInputs. I don't have riscv64 to be able to test locally.

siriobalmelli commented:
> maybe we can move foundationdb into `pkgs/by-name` and get rid of the fdbPackages and `foundationdb73`-prefix?

I'd love to, but I don't fully understand why there are separate `foundationdb/cmake.nix` and `foundationdb/python.nix`. I'm by no means opposed to the idea though, feel free to add to the commit series.

onny commented on nixos/tests/all-tests.nix:
```diff
@@ -439,6 +439,7 @@ in {
   fontconfig-default-fonts = handleTest ./fontconfig-default-fonts.nix {};
   forgejo = handleTest ./forgejo.nix { forgejoPackage = pkgs.forgejo; };
   forgejo-lts = handleTest ./forgejo.nix { forgejoPackage = pkgs.forgejo-lts; };
+  foundationdb = handleTest ./foundationdb.nix {};
```
Please migrate to runTest https://github.com/NixOS/nixpkgs/issues/386873

siriobalmelli replied:
Done. Unable to run and verify at the moment, but anyways until we fix the underlying foundationdb coredump issue it's a moot point.

onny commented:
here's my cleanup by moving foundationdb to pkgs/by-name https://github.com/NixOS/nixpkgs/pull/396805

siriobalmelli committed:
nixos/foundationdb: fix systemd service issues

Taken from https://github.com/NixOS/nixpkgs/issues/378888

- The way the pidfile/lockfile option is coded creates 2 connected issues:
    - 'fdb' needs write permission to the parent of the file to be able to recreate it
      when needed but it is not done in the module
    - The PIDFile option in systemd, makes systemd delete the file when the service is stopped,
      which on the next start will make fdb try to create it,
      but won't be able to because it won't have the write permission on the parent dir.
      Also PIDFile is for forking services, not simple ones and it's also deprecated anyway.
- The line 'PermissionsStartOnly = true;' needs to be removed because otherwise
  the cluster file created on the first run is owned by root which then prevents fdb from changing the file when needed,
  like when updating the coordinators or the cluster description.
  Also, I think this option is deprecated too.

Author: Paul Grandperrin <paul.grandperrin@gmail.com>
Signed-off-by: Sirio Balmelli <sirio@b-ad.ch>

siriobalmelli committed:
nixosTests/foundationdb: new test

Signed-off-by: Sirio Balmelli <sirio@b-ad.ch>

onny committed:
python3Packages.foundationdb: init 7.3.42

siriobalmelli committed:
foundationdb: clean up 'with' syntax

Signed-off-by: Sirio Balmelli <sirio@b-ad.ch>
Co-authored-by: Jonas Heinrich <onny@project-insanity.org>

siriobalmelli committed:
foundationdb: 7.3.42 -> 7.3.62

WIP: still segfaults

siriobalmelli commented:
> here's my cleanup by moving foundationdb to pkgs/by-name #396805

@onny unsure how best to handle this.

In case it's useful, I cherry-picked your commit `c4ac0aeb` on top of this commit series as well, but applying `17d08970` is not straightforward.

If #396805 gets merged then I'll need to manually recraft this commit series on top of files which have moved and changed, and that's a lot of work for no result because this package is effectively useless until we debug the coredump issue.

Any ideas on the coredump issue, or advice on how best to work with #396805 is appreciated 🙏

onny commented:
i would be happy id we could merge your work first. if the package was broken and is still broken after this PR we could mark it as broken and still update it?

PaulGrandperrin commented:
I've been using https://github.com/PaulGrandperrin/nix-systems/blob/main/packages/foundationdb-bin/package.nix for a few months, and it works very well.

I use it both for launching the service and linking client code.

Even if the source package is eventually fixed, it can still make sense to have a binary version, just like we have `firefox` and `firefox-bin`.

The only annoying thing is that my current code is a little bit weird because it first builds two packages that mirrors the `client` and `server` deb packages they are built from, and then merge them both to create a package similar to the one already in nixpkgs, so that it's compatible with the module system code.

siriobalmelli commented:
> i would be happy id we could merge your work first. if the package was broken and is still broken after this PR we could mark it as broken and still update it?

Fair enough. I'm going to try updating it again and see if I can trace down this segfault one more time - otherwise we'll mark it broken.

In the meantime, @onny if you rebase on top of this PR I'm happy to cherry-pick from you (as you prefer).

siriobalmelli commented:
I won't be able to follow up on this, closing PR. @onny and @PaulGrandperrin thanks for your input on this. If I can help with one of your PRs on the topic feel free to tag me and I'll do the best I can.
